lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+# and create a KBaseGenomeAnnotations.Assembly/GenomeAnnotation  object/set
+###########################################################################
+
+use strict;
+use warnings;
+
+use Bio::KBase::kbaseenv;
+use Config::Simple;
+use Data::Dumper qw(Dumper);
+use File::Spec::Functions qw(catfile splitpath);
+use File::Path qw(make_path);
+use File::Copy;
+use Carp qw(croak);
+use Bio::SeqIO;
+use Bio::Perl;
+use Bio::Tools::CodonTable;
+use JSON;
+use Encode qw(encode decode);
+use URI::Encode qw(uri_encode uri_decode);
+use File::Basename;
+use Array::Utils qw(:all);
+use Text::Trim qw(trim);
+use Data::Structure::Util qw( unbless );
+
+use Bio::KBase::Exceptions;
+use Bio::KBase::utilities;
+use Bio::KBase::kbaseenv;
+use GenomeTypeObject;
+
+use Bio::KBase::GenomeAnnotation::GenomeAnnotationImpl;
+use installed_clients::GenomeAnnotationAPIClient;
+use installed_clients::GenomeAnnotationClient;
+use installed_clients::AssemblyUtilClient;
+use installed_clients::GenomeFileUtilClient;
+use installed_clients::WorkspaceClient;
+use installed_clients::KBaseReportClient;
+use installed_clients::kb_SetUtilitiesClient;
+
+
+#######################Begin refactor annotate_process########################
+## helper subs
+sub _get_scientific_name_for_NCBI_taxon {
+    my ($self, $tax_id, $timestamp) = @_;
+    my $url = $self->{_re_url} . "/api/v1/query_results?stored_query=ncbi_fetch_taxon";
+    my $content = encode_json({
+        'ts' => $timestamp,
+        'id'=> $tax_id . '' # make sure tax id is a string
+        });
+    my $req = HTTP::Request->new(POST => $url);
+    $req->header('content-type', 'application/json');
+    $req->content($content);
+    my $ua = LWP::UserAgent->new();
+    
+    my $ret = $ua->request($req);
+    if (!$ret->is_success()) {
+        print("Error body from Relation Engine on NCBI taxa query:\n" .
+            $ret->decoded_content({'raise_error' => 1}));
+        # might want to try to parse content to json and extract error, try this for now
+        die "Error contacting Relation Engine: " . $ret->status_line();
+    }
+    my $retjsonref = $ret->decoded_content({'raise_error' => 1, 'ref' => 1});
+    my $retjson = decode_json($retjsonref);
+
+    if (!$retjson->{count}) {
+        die "No result from Relation Engine for NCBI taxonomy ID " . $tax_id;
+    }
+    return $retjson->{'results'}[0]{'scientific_name'};
+}
+
+#
+## get the genome data via GenomeAnnotationAPIClient
+#
+sub _get_genome {
+    my ($self, $gn_ref) = @_;
+    my $ga_client = new installed_clients::GenomeAnnotationAPIClient($self->{call_back_url});
+    my $output = $ga_client->get_genome_v1({
+        genomes => [{
+            "ref" => $gn_ref
+        }],
+        ignore_errors => 1,
+        no_data => 0,
+        no_metadata => 0,
+        downgrade => 0
+    });
+    my $genome = $output->{genomes}->[0];
+    $genome->{data}->{'_reference'} = $genome->{info}->[6]."/".$genome->{info}->[0]."/".$genome->{info}->[4];
```

Comment:
You can omit the arrows when using references after the first one. This might be clearer to read if you added in some spaces around the `. "/" .`

```perl
$genome->{ data }{ _reference } = $genome->{ info }[6] . "/" . $genome->{ info }[0] . "/" . $genome->{ info }[4];
```